Use std::shared_ptr for DistAutogradContext.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29770 Use std::shared_ptr for DistAutogradContext.**

We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

Commit:
Use std::shared_ptr for DistAutogradContext.

We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

rohan-varma commented on test/dist_autograd_test.py:
```diff
@@ -1253,5 +1253,68 @@ def test_backwards_nested_python_udf(self):
+        t.start()
+
+        with self.assertRaisesRegex(RuntimeError, "Could not find autograd context with id"):
+            dist_autograd.backward([t1.sum()])
```
I'm not 100% clear on what behavior is being tested here and what should be different than from before this PR. Does this result in the backward call throwing an exception (if so, what is exactly is the exception that should be thrown)?

pritamdamania87 replied:
The exception being thrown is "Could not find autograd context id for". This happens because  the autograd context is cleaned up and some thread on some node is still looking for that autograd context.

Before this PR, this test would cause the process to crash since a thread would be using a reference to `DistAutogradContext` but our clean up logic would delete the context from the container.

mrshenli replied:
Does this assume that the thread cleared the context before the backward uses it? How do we guarantee that?

pietern commented on test/dist_autograd_test.py:
```diff
@@ -1,6 +1,7 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import time
+import sys
```
Please stick to alphabetical order.

pietern commented on test/dist_autograd_test.py:
```diff
@@ -1253,5 +1253,68 @@ def test_backwards_nested_python_udf(self):
+        # ensures we simulate a case where we clean up the context while the
+        # backward pass is running.
+        while not DistAutogradTest._my_backward_func_executed:
+            time.sleep(0.1)
```
This assumes the backward pass takes longer than 0.1s. Potential flakiness.

pietern commented on torch/csrc/distributed/autograd/functions/recvrpc_backward.h:
```diff
@@ -9,7 +9,9 @@ namespace torch {
 
+// Forward declarations.
 class DistAutogradContext;
+using ContextPtr = std::shared_ptr<DistAutogradContext>;
```
Defined twice? Also in `dist_autograd_container.h`.

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

pietern commented on test/dist_autograd_test.py:
```diff
@@ -1253,5 +1253,69 @@ def test_backwards_nested_python_udf(self):
+        # backward pass is running.
+        with DistAutogradTest._my_backward_func_executed:
+            DistAutogradTest._my_backward_func_executed.wait()
+        dist_autograd._release_context(context._context_id())
```
Is it possible that this executes after the backwards pass has completed? It's a long shot, seeing as this thread would have to be stalled for at least another 50 autograd steps, but calling it out for the sake of robustness.

Would it be possible to simply call `release_context` from the `backward` function itself?

Otherwise, adding another condition to mark completion of the release would make this always work. In the current setting, you could insert a sleep before the `release_context` and make the test fail, which means that unfortunate thread scheduling could in theory make the test fail.

pietern reviewed (commented):
Thanks for addressing, @pritamdamania87. One more pedantic concern.

mrshenli commented on torch/csrc/distributed/autograd/engine/dist_engine.h:
```diff
@@ -67,7 +67,7 @@ class TORCH_API DistEngine {
   // them in outputEdges.
   void computeDependencies(
-      DistAutogradContext& context,
+      ContextPtr context,
```
This can be a reference of the shared_ptr, as the call site always holds a copy of the `shared_ptr`?

pietern replied:
You can use a const ref. A const ref to a `shared_ptr` doesn't mean the wrapped object is const.

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

pritamdamania87 commented:
@pietern @mrshenli Made a few changes to ensure there is no race in the unit test. This was a little tricky since `release_context` clears the local context, but sends async rpcs to clear the context on other nodes. In the updated unit test I've ensured we wait for all contexts to be cleaned up on all nodes before proceeding with the backward pass.

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -1292,5 +1292,77 @@ def test_backwards_nested_python_udf(self):
+        def backward(ctx, input):
+            assert(DistAutogradTest._test_clean_context_backward_context_id is not None)
+
+            # Release the context to simulate error (use barrier before releasing context to ensure all nodes execute the backward function).
```
comment line too long

pritamdamania87 replied:
What is the convention for python line length in OSS? The python linter isn't complaining and I recall internally the line length is 150 chars.

mrshenli replied:
I am not aware if there is any convention here, I personally use 100. Feel free to ignore this comment though if all linters do not complain

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -1292,5 +1292,77 @@ def test_backwards_nested_python_udf(self):
+        # Send the context id to all nodes.
+        for i in range(0, self.world_size):
+            if i != self.rank:
+                rpc.rpc_sync("worker{}".format(i), _set_rpc_done, args=(context_id, 1))
```
why the rank distance is always 1 for _set_rpc_done here?

pritamdamania87 replied:
Didn't matter for this test case, will fix it though.

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -1292,5 +1292,77 @@ def test_backwards_nested_python_udf(self):
+        for i in range(0, 100):
+            dst = self._next_rank()
+            t1 = rpc.rpc_sync("worker{}".format(dst), torch.add, args=(t1, t1))
+            if i == 99:
```
maybe just do this outside of this loop instead of having a if clause here?

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -1292,5 +1292,77 @@ def test_backwards_nested_python_udf(self):
+
+            # Release the context to simulate error (use barrier before releasing context to ensure all nodes execute the backward function).
+            dist.barrier()
+            dist_autograd._release_context(DistAutogradTest._test_clean_context_backward_context_id)
```
Is this blocking? It has to be to guarantee correctness, right?

pritamdamania87 replied:
This is not blocking, it releases the local context and then just sends async RPCs to release other contexts. The method below `_all_contexts_cleaned_up` is blocking and ensures that contexts are cleaned up on all nodes.

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -1292,5 +1292,77 @@ def test_backwards_nested_python_udf(self):
+            dist_autograd.backward([t1.sum()])
+
+        # HACK: Killing workers since otherwise the autograd engine gets stuck on
+        # other nodes. The proper fix would be addressing:
```
Let's be more specific on why it might stuck on others. 

Would I be correct if I assume it stuck because the crashing backward destroyed the context on this node, and hence the `next_rank` won't be able to clear the context when exiting the scope.

mrshenli commented on torch/csrc/distributed/autograd/functions/recvrpc_backward.h:
```diff
@@ -20,7 +21,7 @@ class TORCH_API RecvRpcBackward : public torch::autograd::Node {
   explicit RecvRpcBackward(
       const AutogradMetadata& autogradMetadata,
-      DistAutogradContext& autogradContext,
+      std::shared_ptr<DistAutogradContext> autogradContext,
```
Why this is not using `ContextPtr`?

pritamdamania87 replied:
Would have a circular dependency, thats why we need to have a forward declaration for `DistAutogradContext`.

mrshenli commented on torch/csrc/distributed/autograd/functions/recvrpc_backward.h:
```diff
@@ -30,7 +31,7 @@ class TORCH_API RecvRpcBackward : public torch::autograd::Node {
 
   // Hold a reference to the autograd context.
-  DistAutogradContext& autogradContext_;
+  std::shared_ptr<DistAutogradContext> autogradContext_;
```
same here

mrshenli reviewed (commented):
The changes LGTM!

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

Commit:
Update on "Use std::shared_ptr for DistAutogradContext."


We were passing around const and non-const references for
DistAutogradContext from DistAutogradContainer. This wasn't safe since the
context could be deleted from the container and a thread might still be using
the reference. This usually would happen when a backward pass fails on the node
driving the backward pass (resulting in delete context messages being sent to
all nodes) but other nodes are still executing code related to that autograd
context.

This was also the reason why `test_backward_autograd_engine_error` was flaky.

Using a std::shared_ptr everywhere ensures we're safe and never crash.

Closes #28928
Closes #26922

Differential Revision: [D18494814](https://our.internmc.facebook.com/intern/diff/D18494814/)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
New commit triggers build failures. Will drop "request changes" when build and test pass.

mrshenli reviewed (approved):
Tests look OK now. Created #30110 to track the failed test, which is irrelevant to this PR.
